### ISSUE TITLE
Reject modifications to constant pointers rather than ignoring them

### ISSUE THREAD

**Issue by haselwarter** [jasmin-lang/jasmin#81]
The below program allocates a variable on the stack, sets it, and passes a pointer to it to a subprocedure, which modifies the pointer contents but does not return it. I would expect this program to be rejected with an error; instead the modification gets eliminated silently and a valid assembly program is emitted.

```
fn foo (reg ptr u64[1] px) {
  px[0] = 1;
}

export fn main () -> reg u64 {
  stack u64[1] x;
  reg u64 res;
  x[0] = 41;
  foo(x);
  res = x[0];
  return res;
}
```

The mutation of the stack seems to disappear in the `unrolling` pass:
```
(* After remove unused function *)


fn foo (Reg const ptr u64[1] px.136) -> () {
  
  px.136[u64 0 ] =(u64) ((64u) 1);
  return ();
}

fn main () -> (Reg u64) {
  Reg u64 res.134;
  Stack u64[1] x.135
  x.135 =(u8[8]) array_init(8);
  x.135[u64 0 ] =(u64) ((64u) 41);
   () = foo(x.135);
  res.134 =(u64) x.135[u64 0 ];
  return (res.134);
}


(* -------------------------------------------------------------------- *)
(* After unrolling *)


fn foo (Reg const ptr u64[1] px.136) -> () {
  
  
  return ();
}

fn main () -> (Reg u64) {
  Reg u64 res.134;
  Stack u64[1] x.135
  x.135 =(u8[8]) array_init(8);
  x.135[u64 0 ] =(u64) ((64u) 41);
   () = foo(x.135);
  res.134 =(u64) x.135[u64 0 ];
  return (res.134);
}


(* -------------------------------------------------------------------- *)
```

**Comment by eponier** [comment]
Thanks for the report! The problem is that the analysis that would detect the invalid write (stack_alloc) is performed after deadcode removal, so at that time there is nothing to detect any more.
Maybe we should add checks at typing time? More generally, there is room for a lot of improvements regarding typing in general and mutable/const in particular.

**Comment by vbgl** [comment]
By design of the jasmin programming language, you can do local reasoning. In the following excerpt, you can prove that `res` will be 41 without looking at the body of `foo`.

~~~
  x[0] = 41;
  foo(x);
  res = x[0];
~~~

There is no way a function can modify the local variables of an other one. Functions communicate only through arguments and returned values and through the external memory.

**Comment by haselwarter** [comment]
Thanks for the explanation Vincent, that was my understanding as well, but it's really at odds with what my programming mind expects from imperative code. So while the code is correct with respect to the functional high level semantics (as one should expect, seeing how it's proven in Coq 😉) it does not follow the principle of least surprise.

Jean-Christophe, yes I was surprised to see that pretyping infers `px` as `const` but doesn't then barf at `px[0] = 1;`. I can't think of a scenario where that's not a programmer error; either `px` should be returned (but the habit of programming in C led the  user to believe the mutation is enough) or it should not be modified. Indeed if later on in `foo` there's a read from `px[0]` the program does get rejected with
```
compilation error in function foo:
stack allocation: cannot write to the constant pointer px (defined at line 1)
```

Rejecting programs early is nice but not crucial for correctness, but this dead code elimination almost felt like obfuscation when I first saw it; unless one is fairly familiar with Jasmin one would expect `main` to return 1 and not 41.

**Comment by vbgl** [comment]
Fixed by #82.
Thanks Philipp for the report and the fix.
